Hibernate - Create user service

flashforward55 opened the pull request.

flashforward55 committed:
test: add comprehensive authentication testing

- Initialize AuthenticationService using Injector
- Test user registration with valid credentials
- Test login with correct credentials
- Test duplicate email registration prevention
- Test login with incorrect password
- Test login with non-existent email
- Add detailed console output for all test cases

flashforward55 committed:
feat: implement user authentication system

Model Layer:
- Add User entity with email, password, and salt fields
- Configure email as unique constraint

DAO Layer:
- Implement UserDao with add and findByEmail methods
- Add HQL query for user lookup by email

Security Layer:
- Add HashUtil with SHA-512 password hashing and salt generation
- Create AuthenticationException and RegistrationException
- Implement UserService with automatic password hashing
- Implement AuthenticationService with login and register

Business Logic:
- Register: Check for duplicate emails before creating user
- Login: Verify credentials using salt and hashed password
- Hash all passwords with unique salt before storage

Testing:
- Add comprehensive Main class with 5 test scenarios
- Test registration, login, duplicates, and invalid credentials

Configuration:
- Add User entity mapping to hibernate.cfg.xml
- Configure database connection and schema generation

This implementation provides secure user authentication with:
- Password hashing using SHA-512
- Unique salt per user
- Duplicate email prevention
- Comprehensive error handling

flashforward55 committed:
feat: add User model and DAO layer

- Create User entity with authentication fields
- Implement UserDao for user persistence
- Add findByEmail method for login lookup

flashforward55 committed:
Implement Hibernate - Create user service

mateacademy-ai-mentor commented on src/main/java/mate/academy/security/AuthenticationServiceImpl.java:
```diff
@@ -0,0 +1,40 @@
+                return user;
+            }
+        }
+        throw new AuthenticationException("Incorrect email or password!");
```
This implementation violates checklist item #7: "Create only one condition for throwing `AuthenticationException` in `login()` method. You may combine two checks: whether the user has been found by login and do passwords match."

While the code is functional, it uses nested `if` statements for the success path. A better approach would be to combine the checks into a single condition to determine if the login should succeed or fail, which makes the code flatter and easier to read. 

Consider refactoring to check for the failure conditions first (e.g., user not found OR password incorrect) and throw the exception, then return the user if the conditions are not met.

mateacademy-ai-mentor commented on src/main/java/mate/academy/service/impl/CinemaHallServiceImpl.java:
```diff
@@ -26,4 +33,34 @@ public CinemaHall get(Long id) {
+            user.setPassword(password);
+            return userService.add(user);
+        }
+    }
```
The `AuthenticationServiceImpl` should be a top-level class in the `mate.academy.security` package. Placing it as a nested static class inside `CinemaHallServiceImpl` is incorrect and leads to code duplication, as there is another implementation in the correct location. This entire nested class should be removed from this file.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Great job on implementing the core authentication functionality! The registration process, password hashing, and DAO layer are all well done. However, there are a couple of critical issues that need to be addressed before this can be approved.

First, there is a complete, duplicate implementation of `AuthenticationServiceImpl` located as a nested class inside `src/main/java/mate/academy/service/impl/CinemaHallServiceImpl.java`. This entire nested class should be removed to maintain a clean project structure and avoid code duplication.

Second, in the correct `AuthenticationServiceImpl.java`, the `login` method uses nested `if` statements. This violates a specific requirement [CHECKLIST ITEM #7] to have only one condition for throwing the `AuthenticationException`. Please refactor this to check for failure conditions first (e.g., `user.isEmpty() || !isPasswordValid(...)`) and throw the exception in a single `if` block.

Please fix these two points and resubmit. Keep up the great work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

flashforward55 committed:
fix: refactor authentication and remove duplicate code

- Use single condition for AuthenticationException in login method
- Combine user existence and password validation checks
- Remove duplicate AuthenticationServiceImpl from CinemaHallServiceImpl
- Extract password validation to separate method for clarity

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Fantastic work on this iteration! I am approving your solution. You've successfully addressed all the feedback from the previous review, and the code is now clean, robust, and fully compliant with all requirements. The refactoring in the `AuthenticationServiceImpl` is excellent, and the overall quality of the implementation is very high. Keep up the great work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
